File: src/test/java/no/nav/tjenestepensjon/simulering/config/CacheConfigTest.java
package no.nav.tjenestepensjon.simulering.config;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.notNullValue;
import static org.hamcrest.Matchers.nullValue;

import static no.nav.tjenestepensjon.simulering.config.TestCacheConfig.SLEEP_TIME;
import static no.nav.tjenestepensjon.simulering.config.TestCacheConfig.TEST_CACHE;
import static no.nav.tjenestepensjon.simulering.config.TestCacheConfig.TEST_CACHE_EXPIRES;

import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.autoconfigure.cache.CacheAutoConfiguration;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.cache.CacheManager;

import no.nav.tjenestepensjon.simulering.config.TestCacheConfig.TestCacheable;

@SpringBootTest(classes = {TestCacheConfig.class, CacheAutoConfiguration.class})
class CacheConfigTest {

    @Autowired
    private CacheManager cacheManager;
    @Autowired
    private TestCacheable testCacheable;

    @Test
    void shouldCacheResponses() {
        String input = "input";
        assertThat(cacheManager.getCache(TEST_CACHE), is(notNullValue()));
        assertThat(cacheManager.getCache(TEST_CACHE).get(input, String.class), is(nullValue()));

        long startTime = System.currentTimeMillis();
        String value = testCacheable.getCacheable(input);
        long endTime = System.currentTimeMillis() - startTime;
        String valueFromCache = cacheManager.getCache(TEST_CACHE).get(input, String.class);

        assertThat(endTime >= SLEEP_TIME, is(true));
        assertThat(valueFromCache, is(equalTo(value)));
        assertThat(valueFromCache, is("cached"));

        startTime = System.currentTimeMillis();
        String cached = testCacheable.getCacheable(input);
        endTime = System.currentTimeMillis() - startTime;
        assertThat(endTime < SLEEP_TIME, is(true));
        assertThat(cached, is("cached"));
    }

    @Test
    void shouldExpire() throws InterruptedException {
        String input = "input";
        assertThat(cacheManager.getCache(TEST_CACHE_EXPIRES).get(input, String.class), is(nullValue()));
        String cached = testCacheable.getCacheableExpires(input);
        assertThat(cacheManager.getCache(TEST_CACHE_EXPIRES).get(input, String.class), is(equalTo(cached)));
        Thread.sleep(SLEEP_TIME);
        assertThat(cacheManager.getCache(TEST_CACHE_EXPIRES).get(input, String.class), is(nullValue()));
    }
}

File: src/test/kotlin/no/nav/tjenestepensjon/simulering/config/CacheConfigTest.kt
package no.nav.tjenestepensjon.simulering.config

import no.nav.tjenestepensjon.simulering.config.CacheConfig.Companion.getCache
import no.nav.tjenestepensjon.simulering.config.CacheConfigTest.TestCache
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.TestConfiguration
import org.springframework.cache.annotation.Cacheable
import org.springframework.cache.annotation.EnableCaching
import org.springframework.context.annotation.Bean
import java.lang.Thread.sleep
import java.time.Duration
import java.time.temporal.ChronoUnit.SECONDS
import kotlin.system.measureTimeMillis

@SpringBootTest(classes = [TestCache::class])
internal class CacheConfigTest {

    @TestConfiguration
    @EnableCaching
    class TestCache {
        @Bean
        fun testCache() = getCache(TEST_CACHE, TEST_CACHE_EXPIRES)

        @Cacheable(value = [TEST_CACHE])
        fun slowValueFetch(key: String) = "cached".also {
            sleep(DELAY)
        }
    }

    @Autowired
    lateinit var testCache: TestCache

    @BeforeEach
    fun emptyCache() {
        sleep(TEST_CACHE_EXPIRES.toMillis())
    }

    @Test
    fun `Should cache responses`() {
        lateinit var old: String
        val uncachedDelay = measureTimeMillis {
            old = testCache.slowValueFetch(input)
        }
        assertNotNull(old)
        assert(DELAY <= uncachedDelay)

        val cachedDelay = measureTimeMillis {
            assertEquals(old, testCache.slowValueFetch(input))
        }
        assert(DELAY > cachedDelay)
    }

    @Test
    fun `Cached responses should expire`() {
        lateinit var old: String
        val uncachedDelay = measureTimeMillis {
            old = testCache.slowValueFetch(input)
        }
        assertNotNull(old)
        assert(DELAY <= uncachedDelay)

        sleep(TEST_CACHE_EXPIRES.toMillis())

        val cachedDelay = measureTimeMillis {
            assertEquals(old, testCache.slowValueFetch(input))
        }
        assert(DELAY <= cachedDelay)
    }

    companion object {
        private const val input = "input"
        internal const val DELAY = 250L
        internal const val TEST_CACHE = "EXPIRING_TEST_CACHE"
        internal val TEST_CACHE_EXPIRES: Duration = SECONDS.duration
    }
}
